rfcs/polyfill_testing.md

```diff
@@ -17,41 +17,67 @@ when developers are determining whether they can rely on it, just as they would
 use this information to know whether they could rely on the feature in the
 browser.
 
-There are a few methods by which we could support this but each of them
-ultimately is some mechanism for running the polyfill script before running
-tests.
+In order to support testing a polyfill, a runtime argument is added which
+when used specifies a single local Javascript file to be injected into
+responses. The contents of the file will be added as a classic script (to ensure
+the polyfill runs before any tests) automatically to responses with a
+`text/html` MIME type after the `doctype` but before the first tag in the
+document.
```

Comment:
Injecting after the doctype will *work* in most cases but relies on error handling if the test uses `<html>` and/or `<head>` start tags. Further, some tests omit the doctype.

I think the following would work and avoid invalid HTML:

1. If there is a match for `/<head(\s+[^>]*)?>/i`, then inject after the first match.
2. Otherwise, if there is a match for `/<html(\s+[^>]*)?>/i`, then inject after the first match.
3. Otherwise, if there is a match for `/<!doctype[^>]*>/i`, then inject after the first match.
4. Otherwise, inject at the start of the response.

In theory, the regexps could match incorrectly if a test does something like `<head data-x=">">`. However, it looks like there are no such tests currently, and there's no reason to use it as the `>` can be escaped as `&gt;` and HTML parser tests are in a [custom format](https://github.com/html5lib/html5lib-tests/tree/master/tree-construction).

```
$ grep -rE "<h(tml|ead)\s+[^>']*'[^'>]+>" .
$ grep -rE '<h(tml|ead)\s+[^>"]*"[^">]+>' .
```

Reply:
> Further, some tests omit the doctype.

A missing doctype is supported by https://github.com/web-platform-tests/wpt/pull/33338 but I will call it out explicitly in the rfc.

> 
> I think the following would work and avoid invalid HTML:
> 
> 1. If there is a match for `/<head(\s+[^>]*)?>/i`, then inject after the first match.
> 2. Otherwise, if there is a match for `/<html(\s+[^>]*)?>/i`, then inject after the first match.
> 3. Otherwise, if there is a match for `/<!doctype[^>]*>/i`, then inject after the first match.
> 4. Otherwise, inject at the start of the response.
> 
> In theory, the regexps could match incorrectly if a test does something like `<head data-x=">">`. However, it looks like there are no such tests currently, and there's no reason to use it as the `>` can be escaped as `&gt;` and HTML parser tests are in a [custom format](https://github.com/html5lib/html5lib-tests/tree/master/tree-construction).
> 
> ```
> $ grep -rE "<h(tml|ead)\s+[^>']*'[^'>]+>" .
> $ grep -rE '<h(tml|ead)\s+[^>"]*"[^">]+>' .
> ```

The PR for this already uses html5lib to parse the dom. I've updated it to keep parsing past the opening html and head tags if they exist over at https://github.com/web-platform-tests/wpt/pull/33338.

I hope this addresses your concerns, thanks for raising them.

Reply:
Sounds good!